src/main/scala/vsys/db/StateMap.scala

```diff
@@ -0,0 +1,127 @@
+package vsys.db
+
+import java.nio.ByteBuffer
+
+import org.h2.mvstore.`type`.DataType
+import org.h2.mvstore.`type`.ObjectDataType
+import org.h2.mvstore.WriteBuffer
+import org.iq80.leveldb.{DB, WriteBatch}
+
+
+class StateMap[K, V](
+  db: DB,
+  stateName: String,
+  keyType: DataType = new ObjectDataType,
+  valueType: DataType = new ObjectDataType) extends SubStorage(db, "states") {
+
+
+  private val StateNameBytes: Array[Byte] = stateName.getBytes(Charset)
+  private val SizeBytes: Array[Byte] = "__size__".getBytes(Charset)
+  private val Prefix: Array[Byte] = makePrefix(StateNameBytes)
+  private val SizeKey: Array[Byte] = makeKey(StateNameBytes, SizeBytes)
+
+  private def getItemBytes(i: scala.Any, iType: DataType): Array[Byte] = {
+
+    val iByteBuffer: WriteBuffer = new WriteBuffer(iType.getMemory(i))
+    iType.write(iByteBuffer, i)
+    iByteBuffer.getBuffer().array()
+
+  }
+
+  def put(key: K, value: V, batchOpt: Option[WriteBatch] = None): V = {
+    var batch: Option[WriteBatch] = batchOpt
+    if (batchOpt.isEmpty) batch = createBatch()
+    val Array(keyBytes, valBytes) = 
+      Array((key, keyType), (value, valueType)).map {case(i, iType) => getItemBytes(i, iType)}
+    put(makeKey(StateNameBytes, keyBytes), valBytes, batch)
+    setSize(sizeAsLong() + 1, batch)
```

Comment:
is it guaranteed that k is not in the map when this function is called?

Reply:
This is my miss. I added commits to fix this. Thank you.